### PR TITLE
Remove unused `@lru_cache` decorator

### PR DESCRIPTION
**Pull request by Fizzadar**
Spotted this working on [something else.](https://github.com/matrix-org/synapse/pull/13573).

Contributed by Nick @ Beeper (@fizzadar).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). 
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by DMRobertson** [comment]
Debian packaging unit tests failed. I wonder why?

```
 Traceback (most recent call last):
  File "/synapse/build/tests/util/caches/test_descriptors.py", line 459, in test_invalidate_cascade
    r = get_awaitable_result(obj.func1("k1", on_invalidate=top_invalidate))
  File "/synapse/build/tests/test_utils/__init__.py", line 40, in get_awaitable_result
    next(i)
  File "/synapse/build/tests/util/caches/test_descriptors.py", line 445, in func1
    return await self.func2(key, on_invalidate=cache_context.invalidate)
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/twisted/internet/defer.py", line 190, in maybeDeferred
    result = f(*args, **kwargs)
  File "/synapse/build/synapse/logging/context.py", line 753, in g
    return run_in_background(f, *args, **kwargs)
  File "/synapse/build/synapse/logging/context.py", line 814, in run_in_background
    res = defer.ensureDeferred(res)
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/twisted/internet/defer.py", line 1129, in ensureDeferred
    return Deferred.fromCoroutine(coro)
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/twisted/internet/defer.py", line 1105, in fromCoroutine
    return _cancellableInlineCallbacks(coro)
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/twisted/internet/defer.py", line 1815, in _cancellableInlineCallbacks
    _inlineCallbacks(None, gen, status)
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/twisted/internet/defer.py", line 1663, in _inlineCallbacks
    status.deferred.callback(getattr(e, "value", None))
  File "/synapse/build/debian/matrix-synapse-py3/opt/venvs/matrix-synapse/lib/python3.10/site-packages/twisted/internet/defer.py", line 660, in callback
    assert not isinstance(result, Deferred)
```

**Comment by DMRobertson** [comment]
Looks unused to me. We do use `LruCache`, but indirectly via `@cached`.

@matrix-org/synapse-core: any reason not to rip this out?

**Comment by clokep** [comment]
> @matrix-org/synapse-core: any reason not to rip this out?

We can always bring it back via git. If it isn't used we should remove it.

**Comment by DMRobertson** [comment]
Consider me sniped. I have no idea what's going on though.

The test that's failing is

https://github.com/matrix-org/synapse/blob/ed5ab472bb0276e96ee5a15b90419ec00ffd3a34/tests/util/caches/test_descriptors.py#L439-L462

But I think it's not failing because we're removing code: it's failing just because we replaced this `@lru_cache` call with an `@cached`: https://github.com/matrix-org/synapse/blob/ed5ab472bb0276e96ee5a15b90419ec00ffd3a34/tests/util/caches/test_descriptors.py#L451-L454

which is just baffling.

**Comment by DMRobertson** [comment]
> The failure is just baffling.

Further bafflement: if I make `func3` an async function and await it in `func2` then the test passes. Does this mean that an `async` `@cached ` function cannot call synchronous `@cached` functions?

**Comment by richvdh** [comment]
For the record, the one and only reference to `@lru_cache` was removed in #13078.

**Comment by clokep** [comment]
> Edit: or maybe `@cached` isn't meant to decorate a synchronous function??? 🤯

I believe it can only be used with async functions.

**Comment by Fizzadar** [comment]
Thank you for picking this up @DMRobertson ❤️!

**Comment by DMRobertson** [comment]
> > Edit: or maybe `@cached` isn't meant to decorate a synchronous function??? exploding_head
> 
> I believe it can only be used with async functions.

Interesting (and not obvious at all). Searching for `@cached.*\n *def` I see only

https://github.com/matrix-org/synapse/blob/0b7830e457359ce651b293c8748bf636973404a9/synapse/storage/databases/main/roommember.py#L1139-L1141

outside of tests. (There are other sync functions in Synapse that are `@cached`, but they all raise `NotImplementedError`.

However there are _many_ examples of synchronous `@cached` functions in tests.

File: tests/util/caches/test_descriptors.py [hunk 1]
```diff
@@ -478,10 +446,10 @@ async def func1(self, key, cache_context):
 
             @cached(cache_context=True)
             async def func2(self, key, cache_context):
-                return self.func3(key, on_invalidate=cache_context.invalidate)
+                return await self.func3(key, on_invalidate=cache_context.invalidate)
```
Comment (erikjohnston) [inline tests/util/caches/test_descriptors.py:449]: I am a bit confused why these changes are needed, given we seem to use `maybeDeferred`:

https://github.com/matrix-org/synapse/blob/19c0e55ef7742d67cff1cb6fb7c3e862b86ea788/synapse/util/caches/descriptors.py

**Review by erikjohnston — APPROVED** [review]

File: tests/util/caches/test_descriptors.py (hunk 1 above)
Comment (DMRobertson) [reply tests/util/caches/test_descriptors.py:449]: I would love to understand it more. Traceback for completeness.

I spent some time staring at this in a debugger and couldn't divine any insight.

```
==============================================================================
Error: 
Traceback (most recent call last):
  File "/home/runner/work/synapse/synapse/tests/util/caches/test_descriptors.py", line 459, in test_invalidate_cascade
    r = get_awaitable_result(obj.func1("k1", on_invalidate=top_invalidate))
  File "/home/runner/work/synapse/synapse/tests/test_utils/__init__.py", line 40, in get_awaitable_result
    next(i)
  File "/home/runner/work/synapse/synapse/tests/util/caches/test_descriptors.py", line 445, in func1
    return await self.func2(key, on_invalidate=cache_context.invalidate)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/runner/work/synapse/synapse/synapse/logging/context.py", line 753, in g
    return run_in_background(f, *args, **kwargs)
  File "/home/runner/work/synapse/synapse/synapse/logging/context.py", line 814, in run_in_background
    res = defer.ensureDeferred(res)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 911, in ensureDeferred
    return _cancellableInlineCallbacks(coro)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
    _inlineCallbacks(None, g, status)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 1421, in _inlineCallbacks
    status.deferred.callback(getattr(e, "value", None))
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 459, in callback
    assert not isinstance(result, Deferred)
builtins.AssertionError: 

tests.util.caches.test_descriptors.DescriptorTestCase.test_invalidate_cascade
-------------------------------------------------------------------------------
```
